File: cauchy/schemes.js
// Schemes - Port from Scala
// Color scheme generation functions

const OneSpoke = 1.0 / 12.0;

function randomR(randomFn) {
    return 0.8 - randomFn() * 0.8;
}

function rgbLuminance(color) {
    const [r, g, b] = color;
    return 0.2126 * r + 0.7152 * g + 0.0722 * b;
}

function sortPaletteByLuminance(colors) {
    return [...colors].sort((a, b) => rgbLuminance(a) - rgbLuminance(b));
}

function monochromatic(radial, randomFn) {
    const radials = [];
    for (let i = 0; i < 5; i++) {
        radials.push(randomR(randomFn));
    }
    radials.sort((a, b) => a - b);
    const palette = radials.map(r => pixToRGBLuminosity({ a: radial.a, r: r }));
    return sortPaletteByLuminance(palette);
}

function analogous(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: radial.a, r: randomR(randomFn) },
        { a: mod1(radial.a + OneSpoke), r: radial.r },
        { a: mod1(radial.a + OneSpoke), r: randomR(randomFn) },
        { a: mod1(radial.a - OneSpoke), r: radial.r }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function triad(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: radial.a, r: randomR(randomFn) },
        { a: mod1(radial.a + 4 * OneSpoke), r: radial.r },
        { a: mod1(radial.a + 4 * OneSpoke), r: randomR(randomFn) },
        { a: mod1(radial.a - 4 * OneSpoke), r: radial.r }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function complementary(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: radial.a, r: mod1(radial.r - 0.1) },
        { a: radial.a, r: mod1(radial.r + 0.1) },
        { a: mod1(radial.a + 6 * OneSpoke), r: radial.r },
        { a: mod1(radial.a + 6 * OneSpoke), r: randomR(randomFn) }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function splitComplementary(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: mod1(radial.a + 5 * OneSpoke), r: radial.r },
        { a: mod1(radial.a + 5 * OneSpoke), r: randomR(randomFn) },
        { a: mod1(radial.a - 5 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 5 * OneSpoke), r: randomR(randomFn) }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function doubleSplitComplementary(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: mod1(radial.a + 1 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 1 * OneSpoke), r: radial.r },
        { a: mod1(radial.a + 5 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 5 * OneSpoke), r: radial.r }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function square(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: radial.a, r: randomR(randomFn) },
        { a: mod1(radial.a + 4 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 4 * OneSpoke), r: radial.r },
        { a: mod1(radial.a + 6 * OneSpoke), r: radial.r }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

function compound(radial, randomFn) {
    const palette = [
        { a: radial.a, r: radial.r },
        { a: mod1(radial.a - 1 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 1 * OneSpoke), r: randomR(randomFn) },
        { a: mod1(radial.a - 5 * OneSpoke), r: radial.r },
        { a: mod1(radial.a - 5 * OneSpoke), r: randomR(randomFn) }
    ].map(rd => pixToRGBLuminosity(rd));
    return sortPaletteByLuminance(palette);
}

const SchemeGenerators = {
    monochromatic,
    analogous,
    triad,
    complementary,
    splitComplementary,
    doubleSplitComplementary,
    square,
    compound
};

function randomScheme(randomFn) {
    const schemes = Object.values(SchemeGenerators);
    const i = Math.floor(randomFn() * schemes.length);
    const radial = { a: randomFn(), r: randomR(randomFn) };
    return schemes[i](radial, randomFn);
}

File: cauchy/examples/colorwheel/Schemes.scala
package org.yuwakisa.flow.colorwheel

import org.yuwakisa.flow.Stuff._
import org.yuwakisa.flow.colorwheel.ColorWheel.{Radial, Rgb, mod1}
import org.yuwakisa.flow.colorwheel.Luminance.lumsort

import scala.util.Random

object Schemes:
  def OneSpoke: Double = 1d / 12d

  def randomR: Double = 0.8 - Random.nextDouble * 0.8

  def monochromatic(rr: Radial): IndexedSeq[Rgb] =
    Seq.fill(5)(randomR)
      .sorted
      .toIndexedSeq
      .map(Radial(rr.a, _))
      .map(ColorWheel.pixToRGBLuminosity)

  def analogous(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(rr.a, randomR),
      Radial(mod1(rr.a + OneSpoke), rr.r),
      Radial(mod1(rr.a + OneSpoke), randomR),
      Radial(mod1(rr.a - OneSpoke), rr.r))
      .map(ColorWheel.pixToRGBLuminosity)

  def triad(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(rr.a, randomR),
      Radial(mod1(rr.a + 4 * OneSpoke), rr.r),
      Radial(mod1(rr.a + 4 * OneSpoke), randomR),
      Radial(mod1(rr.a - 4 * OneSpoke), rr.r))
      .map(ColorWheel.pixToRGBLuminosity)

  def complementary(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(rr.a, mod1(rr.r - 0.1d)),
      Radial(rr.a, mod1(rr.r + 0.1d)),
      Radial(mod1(rr.a + 6 * OneSpoke), rr.r),
      Radial(mod1(rr.a + 6 * OneSpoke), randomR))
      .map(ColorWheel.pixToRGBLuminosity)

  def splitComplementary(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(mod1(rr.a + 5 * OneSpoke), rr.r),
      Radial(mod1(rr.a + 5 * OneSpoke), randomR),
      Radial(mod1(rr.a - 5 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 5 * OneSpoke), randomR))
      .map(ColorWheel.pixToRGBLuminosity)

  def doubleSplitComplementary(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(mod1(rr.a + 1 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 1 * OneSpoke), rr.r),
      Radial(mod1(rr.a + 5 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 5 * OneSpoke), rr.r))
      .map(ColorWheel.pixToRGBLuminosity)

  def square(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(rr.a, randomR),
      Radial(mod1(rr.a + 4 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 4 * OneSpoke), rr.r),
      Radial(mod1(rr.a + 6 * OneSpoke), rr.r))
      .map(ColorWheel.pixToRGBLuminosity)

  def compound(rr: Radial): IndexedSeq[Rgb] =
    IndexedSeq(
      Radial(rr.a, rr.r),
      Radial(mod1(rr.a - 1 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 1 * OneSpoke), randomR),
      Radial(mod1(rr.a - 5 * OneSpoke), rr.r),
      Radial(mod1(rr.a - 5 * OneSpoke), randomR))
      .map(ColorWheel.pixToRGBLuminosity)

  val All = Seq(monochromatic(_),
    analogous(_),
    triad(_),
    complementary(_),
    splitComplementary(_),
    doubleSplitComplementary(_),
    square(_),
    compound(_))

  def random(): IndexedSeq[Rgb] =
    val i = Random.nextInt(All.length)
    All(i)(Radial(Random.nextDouble, Schemes.randomR))
//      .chain(lumsort)
